dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java

```diff
@@ -0,0 +1,181 @@
+    private volatile WSContextService builtInContextService;
+
+    /**
+     * The default context service provided by the concurrent feature
+     */
+    @Reference(target = "(id=DefaultContextService)", cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile WSContextService defaultContextService;
+
+    private final Map<String, WSContextService> namedContextServices = Collections.synchronizedMap(new HashMap<>());
+
+    /**
+     * The liberty global thread pool executor
+     */
+    @Reference(target = "(component.name=com.ibm.ws.threading)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile ExecutorService globalExecutor;
+
+    /**
+     * A scheduled executor which delegates to the global thread pool
+     */
+    @Reference(target = "(deferrable=false)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile ScheduledExecutorService scheduledExecutor;
+
+    @Reference(service = WSContextService.class, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MULTIPLE)
+    protected void addContextService(WSContextService contextService, Map<String, Object> properties) {
+        String name = getContextServiceName(properties);
+        if (name != null) {
+            namedContextServices.put(name, contextService);
+        }
+    }
+
+    protected void updatedContextService(WSContextService contextService, Map<String, Object> properties) {
+        String name = getContextServiceName(properties);
+        if (name == null) {
+            namedContextServices.entrySet().removeIf(e -> e.getValue() == contextService);
+        } else if (namedContextServices.get(name) != contextService) {
+            // If name has changed, remove and re-add
+            namedContextServices.entrySet().removeIf(e -> e.getValue() == contextService);
+            namedContextServices.put(name, contextService);
+        }
+    }
+
+    protected void removeContextService(WSContextService contextService) {
+        namedContextServices.entrySet().removeIf(e -> e.getValue() == contextService);
+    }
+
+    private String getContextServiceName(Map<String, Object> properties) {
+        Object name = properties.get("id");
+        if (name instanceof String) {
+            return (String) name;
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public RMAsyncProvider getAsyncProvider(String contextServiceRef) {
+        return new NamedAsyncProvider(contextServiceRef);
+    }
+
+    @SuppressWarnings("unchecked")
+    private RMContext captureContext(String contextServiceName) {
+        WSContextService namedContextService = null;
+        if (contextServiceName != null) {
+            namedContextService = namedContextServices.get(contextServiceName);
+        }
+
```

Comment:
If a context service is not found for a given name, no error or warning is reported.